projects/cds/src/merchandising/facade/cds-merchandising-user-context.service.ts

```diff
@@ -0,0 +1,76 @@
+import { Injectable } from '@angular/core';
+import {
+  ConverterService,
+  ProductSearchPage,
+  ProductSearchService,
+  RouterState,
+  RoutingService,
+} from '@spartacus/core';
+import { combineLatest, Observable } from 'rxjs';
+import { distinctUntilChanged, map } from 'rxjs/operators';
+import {
+  MERCHANDISING_FACET_NORMALIZER,
+  MERCHANDISING_FACET_TO_QUERYPARAM_NORMALIZER,
+} from '../connectors/strategy/converters';
+import { MerchandisingUserContext } from '../model/merchandising-user-context.model';
+
+function isSameUserContext(
+  previous: MerchandisingUserContext,
+  current: MerchandisingUserContext
+): boolean {
+  return (
+    previous.category === current.category &&
+    previous.productId === current.productId
+  );
+}
+
+@Injectable({
+  providedIn: 'root',
+})
+export class CdsMerchandisingUserContextService {
+  constructor(
+    private routingService: RoutingService,
+    private productSearchService: ProductSearchService,
+    private converterService: ConverterService
+  ) {}
+
+  getUserContext(): Observable<MerchandisingUserContext> {
+    return combineLatest([this.getNavigationContext(), this.getFacets()]).pipe(
+      map(([userContext, facets]: [MerchandisingUserContext, string]) => {
+        const userContextData: MerchandisingUserContext = {
+          ...userContext,
+          facets,
+        };
+        return userContextData;
+      })
+    );
+  }
+
+  private getNavigationContext(): Observable<MerchandisingUserContext> {
+    return this.routingService.getRouterState().pipe(
+      map((routerState: RouterState) => {
+        const productCode = routerState.state.params['productCode'];
+        const categoryCode = routerState.state.params['categoryCode'];
+
+        const routerData: MerchandisingUserContext = {
+          productId: productCode,
+          category: categoryCode,
+        };
+        return routerData;
+      }),
+      distinctUntilChanged(isSameUserContext)
```

Comment:
Could this code maybe help to reduce the complexity of this `distinctUntilChanged`?
```ts


  getUserContext(): Observable<MerchandisingUserContext> {
    return combineLatest([
      this.getProductNavigationContext(),
      this.getCategoryNavigationContext(),
      this.getFacets(),
    ]).pipe(
      map(
        ([productContext, categoryContext, facets]: [
          MerchandisingUserContext,
          MerchandisingUserContext,
          string
        ]) => {
          const userContextData: MerchandisingUserContext = {
            ...productContext,
            ...categoryContext,
            facets,
          };
          return userContextData;
        }
      )
    );
  }

  private getProductNavigationContext(): Observable<MerchandisingUserContext> {
    return this.routingService.getRouterState().pipe(
      map((routerState: RouterState) => {
        return routerState.state.params['productCode'];
      }),
      distinctUntilChanged(),
      map(productCode => {
        const routerData: MerchandisingUserContext = {
          productId: productCode,
        };
        return routerData;
      })
    );
  }

  private getCategoryNavigationContext(): Observable<MerchandisingUserContext> {
    return this.routingService.getRouterState().pipe(
      map((routerState: RouterState) => {
        return routerState.state.params['categoryCode'];
      }),
      distinctUntilChanged(),
      map(categoryCode => {
        const routerData: MerchandisingUserContext = {
          category: categoryCode,
        };
        return routerData;
      })
    );
  }
```

Reply:
Updated to use this approach, but returned strings from each stream rather than `MerchandisingUserContext`

Reply:
Great, even better!